kvscheduler/descriptor/adapter-generator/template.go

```diff
@@ -0,0 +1,365 @@
+type {{ .DescriptorName }}DescriptorBase struct {
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) GetName() string {
+	return "base"
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) KeySelector(key string) bool {
+	return false
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) NBKeyPrefixes() []string {
+	return nil
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) WithMetadata() (withMeta bool, customMapFactory MetadataMapFactory) {
+	return false, nil
+}
+
+{{ if .IsProtoValue -}}
+func (db *{{ .DescriptorName }}DescriptorBase) Build(key string, valueData {{ .ValueDataT }}) (value ProtoValue, err error) {
+{{- if .FromDatasync }}
+	// You can override specific methods of ProtoValue using embedding:
+	//
+	//	type MyProtoValue struct {
+	//		ProtoValue
+	//      typedMsg {{ .ValueDataT }}
+	//	}
+	//	
+	//	// ... (override some methods)
+	//
+	//  return &MyProtoValue{ProtoValue: NewProtoValue(valueData), typedMsg: valueData}	, nil
+	return NewProtoValue(valueData), nil
+{{- else }}
+	return nil, nil
+{{- end }}
+}
+{{- else }}
+func (db *{{ .DescriptorName }}DescriptorBase) Build(key string, valueData {{ .ValueDataT }}) (value {{ .ValueT }}, err error) {
+	return nil, nil
+}
+{{- end }}
+
+func (db *{{ .DescriptorName }}DescriptorBase) Add(key string, value {{ .ValueT }}) (metadata {{ .MetadataT }}, err error) {
+	fmt.Printf("Create for key=%s is not implemented\n", key)
+	return nil, nil
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) Delete(key string, value {{ .ValueT }}, metadata {{ .MetadataT }}) error {
+	fmt.Printf("Delete for key=%s is not implemented\n", key)
+	return nil
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) Modify(key string, oldValue, newValue {{ .ValueT }}, oldMetadata {{ .MetadataT }}) (newMetadata {{ .MetadataT }}, err error) {
+	fmt.Printf("Modify for key=%s is not implemented\n", key)
+	return nil, nil
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) ModifyHasToRecreate(key string, oldValue, newValue {{ .ValueT }}, metadata {{ .MetadataT }}) bool {
+	return false
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) Update(key string, value {{ .ValueT }}, metadata {{ .MetadataT }}) error {
+	fmt.Printf("Modify for key=%s is not implemented\n", key)
+	return nil
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) Dependencies(key string, value {{ .ValueT }}) []Dependency {
+	return nil
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) DerivedValues(key string, value {{ .ValueT }}) []KeyValuePair {
+	return nil
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) Dump(correlate []{{ .DescriptorName }}KVWithMetadata) ([]{{ .DescriptorName }}KVWithMetadata, error) {
+	fmt.Println("Dump is not implemented")
+	return nil, nil
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) DumpDependencies() []string {
+	return nil
+}
+
+////////// Descriptor adapter //////////
+
+type {{ .DescriptorName }}DescriptorAdapter struct {
+	descriptor {{ .DescriptorName }}DescriptorAPI
```

Comment:
By using anonymous field you could omit most of the boiler-plate below and only keep those methods which actually provide some override.

Reply:
good point, will do

Reply:
You mean like this? https://play.golang.org/p/AGDP_nfhisL

Reply:
My terminology is definitely not very correct, because it's not actually override like in C++ classes. They just

https://golang.org/ref/spec#Method_sets
https://golang.org/ref/spec#Selectors

Reply:
cool, will use the anonymous field then